lib/utils/middlewares/context_middleware.ts

```diff
@@ -1,19 +1,32 @@
+import * as express from "express";
+
+import { Option, option } from "ts-option";
 import { right } from "../either";
 
 import { IRequestMiddleware } from "../request_middleware";
 
-import { IContext, IRequestWithContext } from "azure-function-express-cloudify";
+import { IContext } from "azure-function-express";
+
+const CONTEXT_IDENTIFIER = "context";
+
+export function setAppContext(app: express.Express, context: IContext<{}>): void {
+  app.set(CONTEXT_IDENTIFIER, context);
+}
+
+export function getAppContext<T>(request: express.Request): Option<IContext<T>> {
+  return option(request.app.get(CONTEXT_IDENTIFIER));
+}
 
 /**
  * Returns a request middleware that extracts the Azure request context
  * from the request.
  *
  * @param T The type of the bindings found in the context.
- *
- * TODO: validate that the context is indeed defined, respond with ResponseErrorInternal instead
  */
 export function ContextMiddleware<T>(): IRequestMiddleware<never, IContext<T>> {
-  return (request: IRequestWithContext<T>) => {
-    return Promise.resolve(right(request.context));
+  return (request: express.Request) => {
+    const context: IContext<T> = getAppContext<T>(request)
```

Comment:
better return a left with an `IResponseErrorInternal` here or reject the Promise, middlewares are not supposed to throw exceptions.

Reply:
Anyway middleware cannot prevent throwing an exception in case of *unexpected errors* (ie. in called functions), and probably they should throw in this cases. 

While Eithers are good for *expected* errors, I think that every error that causes an 'internal server error' should be an Exception (application should halt). 

Moreover that's exactly the reason why wrapRequestHandler() handles any Exception returned by middlewares returning a ResponseErrorInternal:

https://github.com/teamdigitale/digital-citizenship-functions/blob/cfdbcd03ad78864d0267a380bda5fd88f4006dbf/lib/utils/request_middleware.ts#L14

Reply:
not really, `wrapRequestHandler` handles rejected Promises, it will not catch the exception :)

also, [exceptions should be avoided](https://www.ibm.com/developerworks/library/j-ft13/index.html) since they hijack the execution flow :)

Reply:
According to the output:

wrapRequestHandler|ERROR|http://localhost:7071/api/v1/messages/... |Error: Cannot get context from request

wrapRequestHandler do catch the Exception.

Reply:
nop, just look at the code, there's no `catch`, the exception is catched by [this Promise](https://github.com/teamdigitale/digital-citizenship-functions/blob/master/lib/utils/request_middleware.ts#L135), that is basically the same as rejecting the Promise.

As a matter of style, it is better to use a functional style when handling exception, the control flow is simpler tu understand (no side effect) and it also retain the type of error (that you instead lose with exceptions and rejected promises).